alumnium/cache.py

```diff
@@ -1,34 +1,100 @@
 from os import getcwd
-from typing import Any, Optional, Type
+from typing import Any, Optional
+import json
+import time
+import xxhash
 
 from langchain_core.caches import RETURN_VAL_TYPE, BaseCache
 from langchain_core.load.dump import dumps
 from langchain_core.load.load import loads
-from sqlalchemy import Column, String, create_engine, delete, select
-from sqlalchemy.engine.base import Engine
-from sqlalchemy.orm import Session, declarative_base
+from sqlalchemy import (
+    Column,
+    String,
+    Integer,
+    create_engine,
+    delete,
+    select,
+    ForeignKey,
+    UniqueConstraint,
+)
+from sqlalchemy.orm import Session, declarative_base, relationship
 
 from .logutils import get_logger
 
 logger = get_logger(__name__)
 Base = declarative_base()
 
 
-class Schema(Base):
-    __tablename__ = "alumnium_cache"
-    prompt = Column(String, primary_key=True, nullable=False)
-    llm_string = Column(String, primary_key=True, nullable=False)
+class ModelConfig(Base):
+    __tablename__ = "model_configs"
+    id = Column(Integer, primary_key=True)
+    model_name = Column(String, nullable=False)
+    provider = Column(String, nullable=False)
+    temperature = Column(String, nullable=False)
+    # Add a unique constraint on the combination of fields that make a config unique
+    __table_args__ = (UniqueConstraint("model_name", "provider", "temperature", name="uix_model_config"),)
+
+
+class CacheEntry(Base):
+    __tablename__ = "cache_entries"
+    id = Column(Integer, primary_key=True)
+    model_config_id = Column(Integer, ForeignKey("model_configs.id"))
+    hashed_request = Column(String, nullable=False)
     response = Column(String, nullable=False)
+    created_at = Column(Integer, nullable=False)  # timestamp
+
+    model_config = relationship("ModelConfig")
+
+    __table_args__ = (UniqueConstraint("model_config_id", "hashed_request", name="uix_cache_entry"),)
 
 
 class Cache(BaseCache):
-    def __init__(self, db_path: str = ".alumnium-cache.sqlite", schema: Type[Schema] = Schema):
+    def __init__(self, db_path: str = ".alumnium-cache.sqlite"):
         self.engine = create_engine(f"sqlite:///{getcwd()}/{db_path}")
-        self.schema = schema
-        self.schema.metadata.create_all(self.engine)
-
+        Base.metadata.create_all(self.engine)
         self.session = Session(self.engine)
 
+    def _get_or_create_model_config(self, llm_string: str) -> ModelConfig:
+        config = json.loads(llm_string.split("---")[0])
+        model_config = (
+            self.session.query(ModelConfig)
+            .filter_by(
+                model_name=config["kwargs"]["model_name"],
```

Comment:
Looks like not every Langchain Chat implementation provides this. https://github.com/alumnium-hq/alumnium/actions/runs/15706854101/job/44254631201?pr=118 

You might instead rely on `Model.current` to get a provider and a model name.

Reply:
Thanks for the suggestion! Updated and removed temperature as we don't allow to configure it anyway